### DISCUSSION TITLE
All coverage missing, minimal example

### DISCUSSION THREAD

**Discussion by donmccurdy** [vitest-dev/vitest#5964] [Q&A] [answered]
I'm trying to use Vitest to report code coverage with a minimum viable library. The library is set up as shown below:

```plaintext
├── dist
|  ├── fibonacci.cjs
|  ├── fibonacci.cjs.map
|  ├── fibonacci.d.ts
|  ├── fibonacci.modern.js
|  └── fibonacci.modern.js.map
├── package.json
├── src
|  └── fibonacci.ts
├── test
|  └── fibonacci.test.ts
├── vitest.config.ts
└── yarn.lock
```

[fibonacci_v01.zip](https://github.com/user-attachments/files/15957572/fibonacci_v01.zip)

Note that `package.json` points to outputs in the `dist/` folder, with sourcemaps, as defined here:

```jsonc
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/fibonacci.cjs",
      "default": "./dist/fibonacci.modern.js"
    }
  },
```

Tests pass, but coverage is always empty:

```plaintext
yarn coverage
yarn run v1.22.22
$ vitest run --coverage

 RUN  v1.6.0 /Users/donmccurdy/Demos/2024-06-24_vitest_coverage_issue
      Coverage enabled with istanbul

 ✓ test/fibonacci.test.ts (1)
   ✓ fibonacci

 Test Files  1 passed (1)
      Tests  1 passed (1)
   Start at  11:00:39
   Duration  124ms (transform 20ms, setup 0ms, collect 14ms, tests 1ms, environment 0ms, prepare 38ms)

 % Coverage report from istanbul
--------------|---------|----------|---------|---------|-------------------
File          | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
--------------|---------|----------|---------|---------|-------------------
All files     |       0 |        0 |       0 |       0 |
 fibonacci.ts |       0 |        0 |       0 |       0 | 2-8
--------------|---------|----------|---------|---------|-------------------
✨  Done in 0.69s.
```

I'm experiencing the problem in a larger project, but stripped it down as much as possible to the fibonacci sample shown here. Things I've tried, without success:

1. Use `@vitest/coverage-v8`
2. Use `@vitest/coverage-istanbul`
3. Import directly from `../dist/fibonacci.modern.js` in test, rather than from npm package name
4. Other combinations of `coverage.include` and `coverage.all` in `vitest.config.ts`

No errors or warnings, just 0% coverage. Does Vitest coverage support sourcemaps? Possible that this  is related to v8 or istanbul, but given that the issue is happening with both, I thought I'd start opening a discussion here. Thanks!

**Comment by AriPerkkio** [comment]
Could you convert that zip-file into Github repository or use https://vitest.new ?

> Does Vitest coverage support sourcemaps?

Yes, and we have test case for this:
- https://github.com/vitest-dev/vitest/blob/main/test/coverage-test/test/pre-transpiled-source.test.ts
- https://github.com/vitest-dev/vitest/tree/main/test/coverage-test/fixtures/src/pre-transpiled

The test case runs `pre-transpiled.js` and coverage report shows `original.ts`:

![image](https://github.com/vitest-dev/vitest/assets/14806298/3c590537-b4a8-418c-89cd-e2760cd65ed9)

**Reply by donmccurdy** [reply]
Thanks @AriPerkkio! Here are both:

- https://github.com/donmccurdy/2024-06-24_vitest_coverage_issue
- https://stackblitz.com/edit/vitest-dev-vitest-r6x4s4?file=test%2Fbasic.test.ts

The stackblitz is a little different but I think shows the same issue. I notice there that if the tests imports from the `src/` directory, coverage is shown, but imports from `@vitest/example-test` will print empty coverage. It would be preferable for me if the test suite could run against the builds in `dist/`, so that's why I wasn't importing from `src/`.

**Reply by AriPerkkio** [reply, marked as answer]
Thanks for minimal repros, it was easy to debug the root cause using those. 👍 

This is a bug that's related to `dist` name specifically.

https://github.com/vitest-dev/vitest/blob/0eda99de95e74438581bbefbcc37660ed3132478/packages/vitest/src/defaults.ts#L13-L15

Here the `defaultExclude` is a hard coded default in Vitest. The `this.options.exclude` is what user sets in `coverage.exclude`. We should only use `this.options.exclude` here, and add `defaultExclude` to its default values. Then user could explicitly disable ignoring `dist` directory. Right now it's not possible as `defaultExclude` is always used.

https://github.com/vitest-dev/vitest/blob/0eda99de95e74438581bbefbcc37660ed3132478/packages/coverage-v8/src/provider.ts#L133

https://github.com/vitest-dev/vitest/blob/0eda99de95e74438581bbefbcc37660ed3132478/packages/coverage-istanbul/src/provider.ts#L127

So as work-around for now, rename your `dist` to something else. Could you verify if that works in your setup?

**Reply by donmccurdy** [reply]
Thanks @AriPerkkio, much appreciated! I've confirmed that renaming the `dist` folder to something else, like `build`, fixes the issue. For anyone else encountering this, I also needed to put the `build` folder (or whatever it's named, as long as it's not `dist`) in the "include" list for vitest:

```javascript
import {defineConfig} from 'vitest/config';

export default defineConfig({
  test: {
    coverage: {
      provider: 'istanbul',
      include: ['src/**/*.ts', 'build/**/*'],
    },
  },
});

```

**Reply by AriPerkkio** [reply]
Great, let's follow fix in https://github.com/vitest-dev/vitest/issues/5965.
